### PR TITLE
Fix offline Paraformer LFR boundary padding

### PR DESCRIPTION
**Pull request by LauraGPT**
## Summary

- align standard and NPU Paraformer/SenseVoice LFR windows with the FunASR
  frontend
- repeat the first frame on the left, keep `ceil(T / window_shift)` outputs, and repeat the final frame to complete the last window
- move the transform into a unit-tested helper with checked size arithmetic

## Problem

The current implementation only keeps complete windows starting at input frame 0:

```text
output_frames = (input_frames - window_size) / window_shift + 1
```

For the common `window_size=7`, `window_shift=6` setup, this differs from FunASR in both alignment and output length. It drops the tail, returns only one output for seven input frames instead of two, and can read past the input for some utterances shorter than one window because signed integer division still produces one output frame.

The behavior difference was independently reported in [TigreGotico/onnx-asr#21](https://github.com/TigreGotico/onnx-asr/issues/21): the sherpa and FunASR variants produced different transcripts, while native FunASR agreed with the padded variant.

## Scope

This change covers the standard offline Paraformer and ONNX SenseVoice
recognizers. At the maintainer's request, the SenseVoice single-stream,
batched, and FunASR-Nano compatibility entry points now all delegate to the
same tested LFR helper instead of retaining a second unpadded implementation.

It also updates every existing NPU implementation for these models:

- RKNN and QNN: Paraformer and SenseVoice
- AXCL and AXERA: SenseVoice (there is no Paraformer implementation in these
  folders)
- Ascend: Paraformer and SenseVoice

Fixed-shape RKNN/QNN/AXCL/AXERA models use the shared helper and then truncate
or zero-pad to the model's input frame count. Dynamic-shape Ascend models use
the same helper and retain their maximum-duration truncation. Online
Paraformer keeps its separate streaming buffer path.

## Validation

- regression test was observed failing against the old implementation for boundary values and output counts
- `./scripts/check_style_cpplint.sh 1`
- static and shared Release builds
- `ctest --test-dir build-lfr --output-on-failure -j 8`: 19/19 passed
- focused Release LFR suite: 9/9 passed
- focused LFR suite under GCC ASAN + UBSAN: 9/9 passed
- checked `size_t` output allocation and index arithmetic, including invalid
  fixed shapes and oversized dimensions
- bundled `sherpa-onnx-paraformer-zh-small-2024-03-09` sample produced the same transcript before and after the fix
- bundled SenseVoice five-language batch completed successfully; Chinese and
  Cantonese stayed unchanged, and Korean word spacing improved
- 30 ms and 80 ms SenseVoice inputs completed without an out-of-bounds access,
  invalid tensor shape, or crash

The local validation host does not have the vendor NPU SDKs, so
backend-specific linking and device runtime remain with the SDK-enabled CI and
device environments.

The bundled INT8 SenseVoice model is sensitive to frontend and batch shape:
with the corrected alignment, the single English sample's final token changes
from `gold` to `code`, while the same sample in the five-file batch remains
`gold`. I reproduced this separately with the ONNX model across
`kaldi-native-fbank` and torchaudio features. The shared helper's frame count,
left context, and right padding match FunASR's `apply_lfr`; the previous
transcript came from the old, differently aligned windows.

The commit is SSH-signed and includes a DCO sign-off.

## Summary by CodeRabbit

* **New Features**
  * Added low-frame-rate feature processing with configurable context windows and frame shifts.
  * Added edge padding to preserve complete windows at input boundaries.
  * Added fixed-size output support with zero-padding or truncation.
  * Applied consistent frame processing across supported recognition workflows.

* **Bug Fixes**
  * Improved handling of empty and single-frame inputs.
  * Ensured consistent output frame counts and feature dimensions.
  * Added validation for invalid dimensions, parameters, and input sizes.

### CODE REVIEW

File: sherpa-onnx/csrc/lfr.cc [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace sherpa_onnx {
+
+std::vector<float> ApplyLfr(const std::vector<float> &input,
+                            int32_t input_dim, int32_t window_size,
+                            int32_t window_shift) {
+  assert(input_dim > 0);
+  assert(window_size > 0);
+  assert(window_shift > 0);
+  assert(input.size() % input_dim == 0);
```
[the remark this answers is not in the document]
Comment (LauraGPT) [reply sherpa-onnx/csrc/lfr.cc]: Fixed in `659d4ab`. `ApplyLfr` now validates all three positive dimensions and input-size divisibility before any arithmetic, using the project runtime path (`SHERPA_ONNX_LOGE` + `SHERPA_ONNX_EXIT`) so the checks remain active in Release builds.

Added Release-mode death tests for all four invalid-input cases. Validation: focused LFR test passes, full CTest 19/19 passes, cpplint passes, and the valid path passes under ASAN+UBSAN.

**Comment by csukuangfj** [comment]
Thank you for your contribution! Can you also fix the LFR boundary padding for  sense voice?

**Comment by LauraGPT** [comment]
Done in `a15f171f`. The standard ONNX SenseVoice recognizer now uses the
shared FunASR-compatible `ApplyLfr` helper for single-stream, batched, and
FunASR-Nano compatibility paths; the duplicate unpadded implementation was
removed.

Validation on the exact pushed tree:

- static and shared Release builds passed
- 19/19 CTest targets passed
- 6/6 focused LFR tests passed under ASAN + UBSAN
- all five bundled SenseVoice language samples completed in one batch
- 30 ms and 80 ms inputs completed without invalid shapes or crashes
- cpplint and `git diff --check` passed

I also documented the INT8 model's frontend sensitivity in the PR body: the
single English sample changes its last token from `gold` to `code`, while the
same sample in the five-file batch remains `gold`. I independently compared
the old and corrected transforms with both kaldi-native-fbank and torchaudio;
the new frame count and boundary padding match FunASR's `apply_lfr`.

File: sherpa-onnx/csrc/lfr.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// sherpa-onnx/csrc/lfr.h
+//
+// Copyright (c)  2026  Xiaomi Corporation
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/lfr.h]: Can you update it to use your own information?

File: sherpa-onnx/csrc/lfr.cc [hunk 3]
```diff
@@ -0,0 +1,67 @@
+// sherpa-onnx/csrc/lfr.cc
+//
+// Copyright (c)  2026  Xiaomi Corporation
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/lfr.cc]: Please use your own info.

**Review by csukuangfj — COMMENTED** [review]
Thanks!

Can you also change the paraformer and sense voice for NPUs, e.g., the folders in
sherpa-onnx/csrc/{rknn, qnn, axcl, axera}, etc?

**Comment by LauraGPT** [comment]
Implemented in `ccc65b40c8b8e1f1d8e926a0f8f9cbc0e2ebaff2`.

The shared FunASR-compatible LFR path now covers every existing NPU implementation for these models:

- RKNN: Paraformer and SenseVoice
- QNN: Paraformer and SenseVoice
- AXCL / AXERA: SenseVoice (there is no Paraformer implementation in those folders)
- Ascend: Paraformer and SenseVoice

RKNN/QNN/AXCL/AXERA keep their fixed model shapes by truncating or zero-padding the shared LFR result. Ascend keeps dynamic input shapes and its maximum-duration truncation. The shared helper also uses checked `size_t` allocation/index arithmetic to address the overflow review finding.

Exact-head local validation:

- cpplint passed for all 11 changed C++ files
- focused Release LFR tests: 9/9 passed
- GCC ASAN + UBSAN LFR tests: 9/9 passed
- full Release CTest: 19/19 passed
- static and shared Release builds passed
- Paraformer and SenseVoice official model samples passed, including 30 ms and 80 ms SenseVoice inputs

The validation host does not have the vendor NPU SDKs, so I could not run device-specific QNN/AXCL/AXERA/Ascend linking or inference locally. The exact-head RKNN shared ON/OFF CI jobs are now running. I also updated the PR description with the complete backend scope and validation details.

Follow-up: `cd60df26ab65da8632d2d6b9c7c99e34520bfdd1` moves the output-dimension capacity check before multiplication, so the guard cannot itself wrap on 32-bit `size_t` targets. The same cpplint, Release 9/9, GCC ASAN+UBSAN 9/9, CTest 19/19, and static/shared build gates pass on this exact head.

File: sherpa-onnx/csrc/lfr.h (hunk 2 above)
Comment (LauraGPT) [reply sherpa-onnx/csrc/lfr.h]: Updated in 57f19232. The new LFR file now uses my contributor information (zhifu gao) in the copyright line.

File: sherpa-onnx/csrc/lfr.cc (hunk 3 above)
Comment (LauraGPT) [reply sherpa-onnx/csrc/lfr.cc]: Updated in 57f19232. The new LFR file now uses my contributor information (zhifu gao) in the copyright line.

**Comment by LauraGPT** [comment]
Follow-up pushed in 57f19232:

- changed the two new LFR file copyright lines to my contributor information (zhifu gao), as requested
- made Ascend SenseVoice use explicit fixed LFR constants (input dim 80, window 7, shift 6, output dim 560) throughout preallocation, shape construction, and LFR processing

The CodeRabbit warning that the metadata defaults were 0/0 was not accurate; OfflineSenseVoiceModelMetaData defaults to 7/6. However, the Ascend backend does not read model metadata, so removing that implicit dependency is still the safer and clearer implementation.

Exact-head validation:

- cpplint passed for all three changed files
- Release LFR tests: 9/9 passed
- GCC ASAN + UBSAN LFR tests: 9/9 passed
- full static CTest: 19/19 passed
- full static and shared builds passed
- git diff --check passed

The local host still lacks the Ascend vendor SDK, so the exact Ascend compilation remains delegated to the restarted platform CI.

**Comment by LauraGPT** [comment]
The refreshed 50-job matrix completed with 49 successes and one Windows 11 ARM/Python 3.14 failure. That job built and exercised the package successfully through the full ASR/TTS sequence, then segfaulted in the unrelated Google MedASR example at .github/scripts/test-python.sh:118 while decoding the second 24 kHz input: https://github.com/k2-fsa/sherpa-onnx/actions/runs/30902827158/job/91970988266

The preceding master run of the same Windows ARM job completed successfully: https://github.com/k2-fsa/sherpa-onnx/actions/runs/30810388850/job/91675491946. None of the Paraformer/SenseVoice LFR paths changed by this PR participates in MedASR decoding, and the other 49 current-head jobs pass.

GitHub does not allow the PR author to rerun the upstream job (Must have admin rights). Could a maintainer please rerun this failed job? I will investigate immediately if it reproduces; there is no evidence-based source change to make from this isolated failure.

**Comment by csukuangfj** [comment]
please ignore the CI failures.

**Review by csukuangfj — APPROVED** [review]
Thank you for your contribution!
